File: src/main/java/com/ragin/bdd/cucumber/core/ScenarioStateContext.java
package com.ragin.bdd.cucumber.core;

import static net.javacrumbs.jsonunit.core.Option.IGNORING_ARRAY_ORDER;
import static net.javacrumbs.jsonunit.core.Option.IGNORING_EXTRA_ARRAY_ITEMS;
import static net.javacrumbs.jsonunit.core.Option.IGNORING_EXTRA_FIELDS;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.Getter;
import lombok.Setter;
import net.javacrumbs.jsonunit.core.Option;
import org.springframework.http.ResponseEntity;

@Getter
@Setter
public class ScenarioStateContext {
    private static ScenarioStateContext scenarioStateContext;
    private ResponseEntity<String> latestResponse = null;
    private String fileBasePath = "";
    private String urlBasePath = "";
    private String bearerToken = "";
    private String editableBody = "";
    private String uriPath = "";
    private String defaultBearerToken = "";
    private Map<String, String> headerValues = new HashMap<>();
    private Map<String, String> scenarioContextMap = new HashMap<>();
    private Map<String, String> userTokenMap = new HashMap<>();
    private List<Option> jsonPathOptions = new ArrayList<>(0);

    private ScenarioStateContext() {}

    public static ScenarioStateContext current() {
        if (scenarioStateContext == null) {
            scenarioStateContext = new ScenarioStateContext();
        }
        return scenarioStateContext;
    }

    /**
     * Add IGNORING_EXTRA_ARRAY_ITEMS option to the jsonPathOptions
     */
    public void addJsonIgnoringExtraArrayElements() {
        jsonPathOptions.add(IGNORING_EXTRA_ARRAY_ITEMS);
    }

    /**
     * Add IGNORING_EXTRA_FIELDS option to the jsonPathOptions
     */
    public void addJsonIgnoringExtraFields() {
        jsonPathOptions.add(IGNORING_EXTRA_FIELDS);
    }

    /**
     * Add IGNORING_ARRAY_ORDER option to the jsonPathOptions
     */
    public void addJsonIgnoringArrayOrder() {
        jsonPathOptions.add(IGNORING_ARRAY_ORDER);
    }

    /**
     * Reset states
     */
    public void reset() {
        latestResponse = null;
        fileBasePath = "";
        urlBasePath = "";
        editableBody = "";
        headerValues = new HashMap<>();
        jsonPathOptions = new ArrayList<>(0);
        bearerToken = defaultBearerToken;
    }
}

File: src/main/kotlin/com/ragin/bdd/cucumber/core/ScenarioStateContext.kt
package com.ragin.bdd.cucumber.core

import net.javacrumbs.jsonunit.core.Option
import org.springframework.http.ResponseEntity

object ScenarioStateContext {
    var latestResponse: ResponseEntity<String>? = null
    var fileBasePath: String = ""
    var urlBasePath: String = ""
    var bearerToken: String? = null
    var editableBody: String? = ""
    var uriPath:String = ""
    var defaultBearerToken: String = ""
    var headerValues: HashMap<String, String> = hashMapOf()
    var scenarioContextMap: HashMap<String, String> = hashMapOf()
    var scenarioContextFileMap: HashMap<String, ByteArray> = hashMapOf()
    var userTokenMap: HashMap<String, String> = hashMapOf()
    private var jsonPathOptions: MutableList<Option> = mutableListOf()
    var executionTime = -1L
    var dynamicProxyHost: String? = null
    var dynamicProxyPort: Int? = null
    var polling = Polling()

    /**
     * Add IGNORING_EXTRA_ARRAY_ITEMS option to the jsonPathOptions
     */
    fun addJsonIgnoringExtraArrayElements() {
        jsonPathOptions.add(element = Option.IGNORING_EXTRA_ARRAY_ITEMS)
    }

    /**
     * Add IGNORING_EXTRA_FIELDS option to the jsonPathOptions
     */
    fun addJsonIgnoringExtraFields() {
        jsonPathOptions.add(element = Option.IGNORING_EXTRA_FIELDS)
    }

    /**
     * Add IGNORING_ARRAY_ORDER option to the jsonPathOptions
     */
    fun addJsonIgnoringArrayOrder() {
        jsonPathOptions.add(element = Option.IGNORING_ARRAY_ORDER)
    }

    /**
     * Reset states
     */
    fun reset() {
        executionTime = System.currentTimeMillis()
        latestResponse = null
        fileBasePath = ""
        urlBasePath = ""
        editableBody = ""
        headerValues = hashMapOf()
        jsonPathOptions = mutableListOf()
        bearerToken = defaultBearerToken
        polling = Polling()
        scenarioContextFileMap = hashMapOf()
    }

    fun resolveEntry(key: String): String {
        return scenarioContextMap[key] ?: key
    }

    fun resolveFileEntry(key: String): ByteArray {
        val value = scenarioContextFileMap[key]
        require(value != null) {"Entry $key not found in scendario file context"}

        return value
    }

    fun getJsonPathOptions(): List<Option> {
        return jsonPathOptions
    }

    fun setJsonPathOptions(jsonPathOptions: MutableList<Option>) {
        this.jsonPathOptions = jsonPathOptions
    }

    @JvmStatic
    fun current(): ScenarioStateContext {
        return this
    }

    class Polling {
        var pollEverySeconds: Long = 0
        var numberOfPolls: Int = -1
    }
}
